File: ios/Classes/Bridge/Base/CASMappedChannel.swift
//
//  CASMappedChannel.swift
//  clever_ads_solutions
//
//  Copyright © 2024 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import Flutter

class CASMappedChannel<T> {
    private var map: [String: T] = [:]
    var channelName: String
    private var channel: FlutterMethodChannel

    init(with registrar: FlutterPluginRegistrar, on channelName: String) {
        self.channelName = channelName
        let binaryMessenger = registrar.messenger()
        channel = FlutterMethodChannel(name: "cleveradssolutions/" + channelName, binaryMessenger: binaryMessenger)
        channel.setMethodCallHandler { [weak self] call, result in
            self?.onMethodCall(call, result)
        }
    }

    final func onMethodCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard let id: String = call.getArgAndCheckNil("id", result) else { return }
        switch call.method {
        case "init":
            map[id] = initInstance(id)
            result(nil)
        case "dispose":
            map.removeValue(forKey: id)
            result(nil)
        default:
            guard let instance = map[id] else {
                return result(call.errorFieldNil("id"))
            }
            onMethodCall(instance, call, result)
        }
    }

    open func onMethodCall(_ instance: T, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        print("\(casLogTag) Unknown method '\(call.method)' in '\(channelName)'")
        result(FlutterMethodNotImplemented)
    }

    open func initInstance(_ id: String) -> T {
        fatalError("CASMappedChannel<T>.initInstance(id:) must be overridden in a subclass!")
    }

    subscript(_ id: String) -> T? {
        get {
            return map[id]
        }
        set(newValue) {
            map[id] = newValue
        }
    }

    func remove(_ id: String) -> T? {
        return map.removeValue(forKey: id)
    }

    func invokeMethod(_ id: String, _ methodName: String) {
        channel.invokeMethod(methodName, arguments: ["id": id], result: nil)
    }

    func invokeMethod(_ id: String, _ methodName: String, _ args: [String: Any?]) {
        var newArgs = args
        newArgs["id"] = id
        channel.invokeMethod(methodName, arguments: newArgs, result: nil)
    }
}

File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/CASMappedChannel.kt
package com.cleveradssolutions.plugin.flutter

import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

abstract class CASMappedChannel<T>(
    binding: FlutterPlugin.FlutterPluginBinding,
    private val name: String
) : MethodChannel.MethodCallHandler {
    private val channel = MethodChannel(binding.binaryMessenger, name)
    private val map = HashMap<String, T>()

    init {
        @Suppress("LeakingThis")
        channel.setMethodCallHandler(this)
    }

    final override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val id = call.getArgAndCheckNull<String>("id", result) ?: return
        when (call.method) {
            "init" -> {
                map[id] = initInstance(id)
                result.success()
            }

            "dispose" -> {
                map.remove(id)
                result.success()
            }

            else -> {
                val instance = map[id] ?: return result.errorFieldNull(call, "$id in $this")
                onMethodCall(instance, call, result)
            }
        }
    }

    open fun onMethodCall(instance: T, call: MethodCall, result: MethodChannel.Result) {
        Log.e(CAS_LOG_TAG, "Unknown method '${call.method}' in '$name'")
        result.notImplemented()
    }

    open fun initInstance(id: String): T =
        throw NotImplementedError("MappedMethodHandler<T>.initInstance(id:) must be overridden in a subclass!")

    operator fun get(id: String): T? = map[id]

    operator fun set(id: String, value: T) {
        map[id] = value
    }

    fun remove(id: String) = map.remove(id)

    fun invokeMethod(
        id: String,
        methodName: String,
        args: HashMap<String, Any?>? = null,
        callback: MethodChannel.Result? = null
    ) {
        val map = args ?: HashMap()
        map["id"] = id
        channel.invokeMethod(methodName, map, callback)
    }
}
